### PR TITLE
server/swap/test: fix intermittent TestMalformedSwap failures

### PR DESCRIPTION
**Pull request by LasTshaMAN**
I also pulled out problematic part of `TestMalformedSwap` test into separate `TestRetriesDuringSwap` since it doesn't seem to be related to "testing malformed" (and I've discovered quite a lot of things this way about this functionality).

Closes https://github.com/decred/dcrdex/issues/1730.

### CODE REVIEW

File: server/swap/swap.go [hunk 1]
```diff
@@ -2122,7 +2119,7 @@ func (s *Swapper) processMatchAcks(user account.AccountID, msg *msgjson.Message,
 	}
 	if len(matches) != len(acks) {
 		s.respondError(msg.ID, user, msgjson.AckCountError,
-			fmt.Sprintf("expected %d acknowledgements, got %d", len(acks), len(matches)))
+			fmt.Sprintf("expected %d acknowledgements, got %d", len(matches), len(acks)))
```
Comment (LasTshaMAN) [inline server/swap/swap.go:2122]: Looks like these 2 args should be switched

File: server/swap/swap_test.go [hunk 2]
```diff
@@ -747,9 +748,25 @@ func (rig *testRig) checkMatchNotification(msg *msgjson.Message, oid order.Order
...
+// helper to check the swap status for specified user.
+func (rig *testRig) checkSwapStatusChange(wantStatus order.MatchStatus, user *tUser) error {
+	// Swap counterparty ultimately gets an audit request. Synchronize with that.
+	if err := rig.waitChans("counterparty received our swap -> counterparty audit request", nil, rig.auth.swapReceived, rig.auth.auditReq); err != nil {
+		return err
+	}
+	tracker := rig.getTracker()
+	tracker.mtx.RLock()
+	status := tracker.Status
+	tracker.mtx.RUnlock()
```
Comment (LasTshaMAN) [inline server/swap/swap_test.go:760]: Seems like there are a bunch of places in these tests where mutex should be used, like this one,
if reviewers agree, I'd add the required mutex calls I've spotted in this test-file

File: server/swap/swap.go [hunk 3]
```diff
@@ -983,12 +983,11 @@ func (s *Swapper) processBlock(ctx context.Context, block *blockNotification) {
...
 
-		// Lock the matchTracker so the following checks and updates are atomic
-		// with respect to Status.
 		match.mtx.RLock()
-		defer match.mtx.RUnlock()
+		status := match.Status
+		match.mtx.RUnlock()
 
-		switch match.Status {
+		switch status {
```
Comment (LasTshaMAN) [inline server/swap/swap.go:990]: Does it make sense to use `defer` here ? my linter complains about it, and it seems that `match.Status` is all that we really need to access in this `for` loop (based on `matchTracker.mtx` comment), so I've trimmed down mutex usage in this case,

unless there is another reason to hold this mutex until this function terminates ?

File: server/swap/swap_test.go (hunk 2 above)
Comment (LasTshaMAN) [reply server/swap/swap_test.go:760]: and I've actually encountered a race couple of times while running these tests locally:

```
==================
WARNING: DATA RACE
Write at 0x00c0004be348 by goroutine 38:
  decred.org/dcrdex/server/swap.(*Swapper).processInit()
      /Users/iurii/dcrdex/server/swap/swap.go:1618 +0x1c07
  decred.org/dcrdex/server/swap.(*Swapper).handleInit.func1()
      /Users/iurii/dcrdex/server/swap/swap.go:1957 +0x6b
  decred.org/dcrdex/dex/wait.(*TaperingTickerQueue).Run.func1()
      /Users/iurii/dcrdex/dex/wait/queue.go:197 +0xfa
  decred.org/dcrdex/dex/wait.(*TaperingTickerQueue).Run·dwrap·5()
      /Users/iurii/dcrdex/dex/wait/queue.go:248 +0x47

Previous read at 0x00c0004be348 by goroutine 91:
  decred.org/dcrdex/server/swap.(*testRig).sendSwap_maker()
      /Users/iurii/dcrdex/server/swap/swap_test.go:777 +0x384
  decred.org/dcrdex/server/swap.TestRetrySwapInit()
      /Users/iurii/dcrdex/server/swap/swap_test.go:2041 +0x2de
  testing.tRunner()
      /Users/iurii/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /Users/iurii/go/src/testing/testing.go:1306 +0x47
```

File: server/swap/swap_test.go [hunk 4]
```diff
@@ -823,11 +816,13 @@ func (rig *testRig) sendSwap(user *tUser, oid order.OrderID, recipient string) (
 	} else {
 		rig.abcNode.setContract(swap.coin, false)
 	}
+	rig.auth.mtx.Lock()
 	rig.auth.swapID = swap.req.ID
+	rig.auth.mtx.Unlock()
```
Comment (LasTshaMAN) [inline server/swap/swap_test.go:821]: Also seems like there is a race with `TAuthManager.swapID` field here (and this ^ should fix it):
```
WARNING: DATA RACE
Write at 0x00c0003c1cd0 by goroutine 97:
  decred.org/dcrdex/server/swap.(*testRig).sendSwap()
      /Users/iurii/dcrdex/server/swap/swap_test.go:829 +0x2f9
  decred.org/dcrdex/server/swap.(*testRig).sendSwap_maker()
      /Users/iurii/dcrdex/server/swap/swap_test.go:765 +0x127
  decred.org/dcrdex/server/swap.TestRetrySwapInit()
      /Users/iurii/dcrdex/server/swap/swap_test.go:2045 +0x2de
  testing.tRunner()
      /Users/iurii/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /Users/iurii/go/src/testing/testing.go:1306 +0x47

Previous read at 0x00c0003c1cd0 by goroutine 42:
  decred.org/dcrdex/server/swap.(*TAuthManager).Send()
      /Users/iurii/dcrdex/server/swap/swap_test.go:159 +0x3ac
  decred.org/dcrdex/server/swap.(*Swapper).respondSuccess()
      /Users/iurii/dcrdex/server/swap/swap.go:1258 +0x1c3
  decred.org/dcrdex/server/swap.(*Swapper).processInit()
      /Users/iurii/dcrdex/server/swap/swap.go:1631 +0x2345
  decred.org/dcrdex/server/swap.(*Swapper).handleInit.func1()
      /Users/iurii/dcrdex/server/swap/swap.go:1956 +0x6b
  decred.org/dcrdex/dex/wait.(*TaperingTickerQueue).Run.func1()
      /Users/iurii/dcrdex/dex/wait/queue.go:197 +0xfa
  decred.org/dcrdex/dex/wait.(*TaperingTickerQueue).Run·dwrap·5()
      /Users/iurii/dcrdex/dex/wait/queue.go:248 +0x47
```

File: server/swap/swap.go (hunk 3 above)
Comment (chappjc) [reply server/swap/swap.go:990]: Just for the reason in the comment, to make sure that any other goroutine viewing `matchTracker.Status` and `swapStatus.swapConfirmed` (i.e. `checkInactionEventBased` and `checkInactionBlockBased`) has a consistent view of those things.

It probably doesn't matter, otherwise what's the point of a separate mutex for the `swapStatus`, but it needs to be examined carefully.

File: server/swap/swap_test.go (hunk 4 above)
Comment (chappjc) [reply server/swap/swap_test.go:821]: Maybe another test already had the same issue, but `TAuthManager.swapID` being modified while there is as swap presently being handled with `handleInit` doesn't make sense.  (If a subsequent `sendSwap` modified it while a previous `handleInit` was already running, then the `Send` for the response on the first one would reference the wrong `swapID`! Thus the test rig won't work right unless the swapID happens to be the same like it is in your test).

In other words, this `swapID` field is just a single value because the test rig isn't designed to work properly with multiple concurrent `handleInit` calls, and all of the prior tests respected that.

I think you're only seeing this race now because of what `TestRetrySwapInit` is trying to test.
OK to locking in `sendSwap`, but I don't think other tests were hitting this race, and any that do including this new one are potentially in error.

File: server/swap/swap_test.go [hunk 5]
```diff
@@ -2003,18 +1998,54 @@ func TestMalformedSwap(t *testing.T) {
...
+	// adequate fee and 0 confs).
+	startWaitingTime := time.Now()
+	for {
+		// We might get a couple of duplicate init request errors here, in that case
+		// we simply retry request later.
+		err := rig.sendSwap_maker(false)
+		if err == nil {
+			// We are done, finally got a retry attempt that worked.
+			ensureNilErr(rig.checkSwapStatusChange(order.MakerSwapCast, matchInfo.maker))
+			return
```
Comment (chappjc) [inline server/swap/swap_test.go:2038]: Currently we are indeed done after this, but I'd prefer to `break` the loop instead and achieve the same outcome just by running to the end of the test function.  It'll allow any coded that may eventually get added after the `for` to work.  Linters will call that out, but `go` will still compile and run the test with unreachable code.

**Review by chappjc — COMMENTED** [review]
Looks pretty good over all.  Thanks for the good clean up and fixing my broken test.

Sorry I just noticed this was draft, but I wanted to see.  :)

File: server/swap/swap_test.go (hunk 2 above)
Comment (LasTshaMAN) [reply server/swap/swap_test.go:760]: Interesting ... haven't thought about increasing test "coverage" this way, makes sense; so I see 2 options then:
- leave other tests as is (maybe I can go over those to see whether I can spot any potential concurrency issues and raise if any), preserving this "coverage", letting race detector do its job for us (and address these, if any, on a need-to basis)
- or (lazy approach I've suggested above) use mutex blindly, basically giving up on this additional "coverage" at the benefit of not having to worry about writing / maintaining precise tests with respect to concurrent access

I can see the benefits of either of these (personally I'm used to option 2) and I'm onboard with whichever you prefer

File: server/market/market.go [hunk 6]
```diff
@@ -876,14 +876,14 @@ func (m *Market) processMatchAcksForCancel(user account.AccountID, msg *msgjson.
...
-			"error parsing match request acknowledgment")
+			fmt.Sprintf("error parsing match request acknowledgment: %v", err))
 		return
 	}
 	// The acknowledgment for both the taker and maker should come from the same user
 	expectedNumAcks := 2
 	if len(acks) != expectedNumAcks {
 		m.respondError(msg.ID, user, msgjson.AckCountError,
-			fmt.Sprintf("expected %d acknowledgements, got %d", len(acks), expectedNumAcks))
+			fmt.Sprintf("expected %d acknowledgements, got %d", expectedNumAcks, len(acks)))
```
Comment (LasTshaMAN) [inline server/market/market.go:886]: This is mixed as well (similar to one I've spotted above)

File: server/market/market.go [hunk 7]
```diff
@@ -1992,8 +1992,9 @@ func (m *Market) handlePreimageResp(msg *msgjson.Message, reqData *piData) {
 	if err != nil {
 		log.Errorf("StorePreimage: %v", err)
 		// Fatal backend error. New swaps will not begin, but pass the preimage
-		// along so it does not appear as a miss to collectPreimages.
-		m.respondError(msg.ID, reqData.ord.User(), msgjson.UnknownMarketError, "internal server error")
+		// along so that it does not appear as a miss to collectPreimages.
+		m.respondError(msg.ID, reqData.ord.User(), msgjson.UnknownMarketError,
+			fmt.Sprintf("internal server error: %v", err))
```
Comment (LasTshaMAN) [inline server/market/market.go:1997]: I'm assuming here there is "nothing to hide" from the client; in other words, this error shouldn't contain sensitive information, thoughts ?

File: server/swap/swap.go [hunk 8]
```diff
@@ -1420,7 +1419,7 @@ func (s *Swapper) processAck(msg *msgjson.Message, acker *messageAcker) {
 	ack := new(msgjson.Acknowledgement)
 	err := msg.UnmarshalResult(ack)
 	if err != nil {
-		s.respondError(msg.ID, acker.user, msgjson.RPCParseError, "error parsing acknowledgment")
+		s.respondError(msg.ID, acker.user, msgjson.RPCParseError, fmt.Sprintf("error parsing acknowledgment: %v", err))
```
Comment (LasTshaMAN) [inline server/swap/swap.go:1422]: Otherwise `err` is swallowed (updated a couple of similar places with the same issue, can find them all by searching `fmt.Sprintf` in this PR):
```
2022-07-27 11:57:10.586 [DBG] TEST: Error going to user 224cbaaafa80bf3bd1ff731590bcbdda5a76f91e60a156994634e91c0000002e, code: 1, msg: error parsing acknowledgment
```

File: server/market/market.go (hunk 7 above)
Comment (LasTshaMAN) [reply server/market/market.go:1997]: Also, seems like `msgjson.UnknownMarketError` code here (and in some other similar places) isn't the best fit ? Thought I'd call this out, and maybe fix.

File: server/swap/swap_test.go (hunk 2 above)
Comment (chappjc) [reply server/swap/swap_test.go:760]: If there's a place where you can reason out a race even if it's not detected, use the mutex.

We're sort of used to the lazy approach in much of the product code, and I can say that it bites us.   It's very important to have a good understanding of access patterns because over using mutexes can be bad for not only performance but it can get you into deadlocks and undesirable mutex contention.

File: server/swap/swap.go [hunk 9]
```diff
@@ -1475,15 +1474,16 @@ func (s *Swapper) processAck(msg *msgjson.Message, acker *messageAcker) {
 	if !acker.isMaker {
 		acker.match.Sigs.TakerRedeem = ack.Sig
 		if err = s.storage.SaveRedeemAckSigB(mktMatch, ack.Sig); err != nil {
-			s.respondError(msg.ID, acker.user, msgjson.UnknownMarketError, "internal server error")
+			s.respondError(msg.ID, acker.user, msgjson.UnknownMarketError,
+				fmt.Sprintf("internal server error: %v", err))
```
Comment (chappjc) [inline server/swap/swap.go:1478]: For any of the errors from an Archivist, I'm inclined to log server-side and omit the actual error from the response to client.  Unless we are explicitly checking the error type (see the `Is*` helpers in `server/db/errors.go`), we should assume the error is an internal failure and that's all the client needs to know.

File: server/swap/swap.go [hunk 10]
```diff
@@ -1502,7 +1502,7 @@ func (s *Swapper) processInit(msg *msgjson.Message, params *msgjson.Init, stepIn
 			stepInfo.match.ID(), actor, params.CoinID, params.Contract, err)
 		actor.status.mtx.RUnlock()
 		s.respondError(msg.ID, actor.user, msgjson.ContractError,
-			"redemption error")
+			fmt.Sprintf("contract error encountered: %v", err))
```
Comment (chappjc) [inline server/swap/swap.go:1505]: Good catch!  That's not a redemption error lol.  In this case I don't see the harm in including the full `err`.

File: server/swap/swap_test.go (hunk 2 above)
Comment (chappjc) [reply server/swap/swap_test.go:760]: On this note, while I think it's good to be acutely aware of what goroutines are running and why it is or isn't safe to access some field, I'm no longer fond of the general approach in our tests of being so deeply involved with internals and implementation details rather than testing the API.  The fact that we have to do this is partly a consequence of the need to stub out connections and databases, but it's also a hint that we've not factored out enough logic into smaller functions so we can test them in isolation, among other things like wanting to reuse structures instead of making fresh ones, doing more repeated teardown/bringup, etc.

It's frustrating to change some small detail in the implementation that does not affect the overall function of the system but ends up completely wrecking a test that expects some irrelevant order of operations or some message that we hack a channel into via a stub, etc.

This is not a comment that you should change anything, just saying that while I always prefer a developer to think hard about access patterns and if a mutex is even needed, I generally do not love our overall testing approach.  This is an evolving view of mine.

File: server/market/market.go (hunk 7 above)
Comment (LasTshaMAN) [reply server/market/market.go:1997]: So, would it be ok to replace `msgjson.UnknownMarketError` (and in similar cases where appropriate) with `msgjson. RPCInternalError` ? and revert back `fmt.Sprintf("internal server error: %v", err)` -> `"internal server error"`,

> The error is likely to be from postgres here, and sending it blindly to the client might leak info about server config

since looks like this is applicable only to "internal" cases calling `respondError` I've modified in this PR

Comment (chappjc) [reply server/market/market.go:1997]: That sounds good.  The other unmarshal errors and similar can still go back to the client though I think.

File: server/swap/swap_test.go (hunk 4 above)
Comment (LasTshaMAN) [reply server/swap/swap_test.go]: Revised this a bit, I think I was observing this race while `TestRetrySwapInit` wasn't "complete", now it can locking on this mutex can be safely removed I think (done).

File: server/swap/swap_test.go (hunk 2 above)
Comment (LasTshaMAN) [reply server/swap/swap_test.go]: I've looked a bit more at how `tracker.Status` is accessed, (in `TestRetrySwapInit`) it gets synchronized on `rig.auth.swapReceived` (or for redeems it would be `rig.auth.redeemReceived`) with the concurrent Go-routine that is running `processInit`, thus locking `tracker.mtx` isn't really necessary (so, I removed locking).
In other tests there seems to be no concurrent access at all.

File: server/swap/swap_test.go [hunk 11]
```diff
@@ -2003,18 +1994,54 @@ func TestMalformedSwap(t *testing.T) {
...
-	// Back to NewlyMatched
 	tracker := rig.getTracker()
+	// We're "rewinding time" back NewlyMatched status to be able to retry sending
+	// the same init request again.
+	tracker.mtx.Lock()
 	tracker.Status = order.NewlyMatched
-	// We're rewinding time to send the same init again. To support this hack,
-	// wait for swapSearching to clear.
-	time.Sleep(200 * time.Millisecond)
+	tracker.mtx.Unlock()
```
Comment (LasTshaMAN) [inline server/swap/swap_test.go:2021]: Here, I still think we need to lock in order for next `processInit` to be able to read this change (it will also lock on this `tracker.mtx` for reading),

still, I couldn't reproduce this race (without locking)

Comment (LasTshaMAN) [reply server/swap/swap_test.go:2021]: Ah well, looking at it afresh I can see why there is no race, calling `s.latencyQ.Wait` establishes "happens after" via [this chan](https://github.com/LasTshaMAN/dcrdex/blob/1201862002cab39d58b5d4366bd6e0e251367341/dex/wait/queue.go#L183-L183),

go-routine that will be working on `processInit` will have to receive on this chan first

File: server/market/market.go (hunk 7 above)
Comment (LasTshaMAN) [reply server/market/market.go]: Done.

**Comment by LasTshaMAN** [comment]
Addressed all the things discussed above, and also added retry-test for redeems (similar to init), hopefully you don't mind slight  code rearrangements in tests.

Ready for another review round now.

File: server/swap/swap.go (hunk 3 above)
Comment (chappjc) [reply server/swap/swap.go:990]: Revert this please and we can get this in.

Comment (LasTshaMAN) [reply server/swap/swap.go]: Done.

**Review by chappjc — APPROVED** [review]
